<?xml version="1.0" ?>
<robot name="airport_robot"
    xmlns:xacro="http://www.ros.org/wiki/xacro">
    <!-- ***********麦克纳姆轮小车模型，更改了激光雷达的位置**************** -->
    <xacro:property name="robot_name" value="$(arg robot_name)"/>
    <xacro:property name="PI" value="3.1415926835897931"/>
    <xacro:property name="material_density" value="10.0"/>


    <!-- base -->
    <xacro:property name="base_length" value="1.2"/>
    <xacro:property name="base_width" value="1.0"/>
    <xacro:property name="base_height" value="0.6"/>
    <xacro:property name="base_collision_height" value="0.6"/>
    <xacro:property name="base_mass" value="${material_density*base_length*base_width*base_height}"/>
    <xacro:property name="base_mass_ixx" value="${(base_width*base_width+base_height*base_height)*base_mass/12.0}"/>
    <xacro:property name="base_mass_iyy" value="${(base_length*base_length+base_height*base_height)*base_mass/12.0}"/>
    <xacro:property name="base_mass_izz" value="${(base_width*base_width+base_width*base_width)*base_mass/12.0}"/>
    <!-- rear tyre -->
    <!-- <xacro:property name="rear_tyre_x" value="${base_length/2.0}"/> -->
    <xacro:property name="rear_tyre_x" value="0.4"/>
    <xacro:property name="rear_tyre_y" value="${base_width/2.0}"/>
    <xacro:property name="rear_tyre_r" value="${base_collision_height/2.0}"/>
    <xacro:property name="rear_tyre_length" value="0.1"/>
    <xacro:property name="rear_tyre_mass" value="20"/>
    <xacro:property name="rear_tyre_mass_ixx" value="0.5"/>
    <xacro:property name="rear_tyre_mass_iyy" value="0.9"/>
    <xacro:property name="rear_tyre_mass_izz" value="0.5"/>
    <!-- front tyre -->
    <!-- <xacro:property name="front_tyre_x" value="${base_length/2.0}"/> -->
    <xacro:property name="front_tyre_x" value="0.4"/>
    <xacro:property name="front_tyre_y" value="${base_width/2.0}"/>
    <xacro:property name="front_tyre_r" value="${base_collision_height/2.0}"/>
    <xacro:property name="front_tyre_length" value="0.1"/>
    <xacro:property name="front_tyre_mass" value="20"/>
    <xacro:property name="front_tyre_mass_ixx" value="0.5"/>
    <xacro:property name="front_tyre_mass_iyy" value="0.9"/>
    <xacro:property name="front_tyre_mass_izz" value="0.5"/>
    <xacro:property name="front_commandtyre_mass_izz" value="0.5"/>
    <!-- steering -->
    <xacro:property name="str_angle" value="0.6"/>
    <xacro:property name="str_length" value="0.01"/>
    <xacro:property name="str_radius" value="0.1"/>
    <xacro:property name="str_mass" value="5"/>
    <xacro:property name="str_mass_ixx" value="0.012"/>
    <xacro:property name="str_mass_iyy" value="0.025"/>
    <xacro:property name="str_mass_izz" value="0.012"/>

    <!-- IMU -->
    <xacro:property name="IMU_x" value="0"/>
    <xacro:property name="IMU_y" value="0"/>
    <xacro:property name="IMU_z" value="1.25"/>

    <!-- LiDAR -->
    <xacro:property name="LiDAR_x" value="0"/>
    <xacro:property name="LiDAR_y" value="0"/>
    <xacro:property name="LiDAR_z" value="1.5"/>

    <!-- velodyne -->
    <xacro:property name="lasers_horizon" value="32"/>
    <xacro:property name="lasers_vertical" value="32"/>
    
    
    <!-- To fix the problem -->
    <!-- The root link base_link has an inertia specified in the URDF, but KDL does not support a root link with an inertia.  As a workaround, you can add an extra dummy link to your URDF. -->
    <link name="odom">
    </link>
    <joint name="odom" type="fixed">
        <parent link="odom"/>
        <child link="base_footprint"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>

    <link name="base_footprint">
    </link>
    <joint name="base_joint" type="fixed">
        <parent link="base_footprint"/>
        <child link="base_link"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>

    <!--Car Body-->
    <link name="base_link">
        <inertial>
            <origin xyz="0 0 ${base_collision_height}" rpy="0 0 0"/>
            <mass value="${base_mass}"/>
            <inertia ixx="${base_mass_ixx}" ixy="0" ixz="0" iyy="${base_mass_iyy}" iyz="0" izz="${base_mass_izz}" />
            <!-- <pose>0 0 -0.1 0 0 0</pose>  -->
        </inertial>
        <visual>
            <origin xyz="0 0 ${base_collision_height}" rpy="0 0 0" />
            <geometry>
                <!-- <mesh filename="/home/shenlu/uprobot.stl" scale="0.001 0.001 0.001"/> -->
                <box size="${base_length} ${base_width} ${base_height}" />
            </geometry>
            <material name="Yellow">
                <color rgba="1.0 1.0 0.0 1.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 ${base_collision_height}" rpy="0 0 0"/>
            <geometry>
                <!-- <mesh filename="/home/shenlu/uprobot.stl" scale="0.001 0.001 0.001"/> -->
                <box size="${base_length} ${base_width} ${base_height}"/>
            </geometry>
        </collision>
    </link>

    <!--定位激光雷达-->
    <xacro:arg name="gpu" default="true"/>
    <xacro:property name="gpu" value="$(arg gpu)" />
    <xacro:arg name="organize_cloud" default="false"/>
    <xacro:property name="organize_cloud" value="$(arg organize_cloud)" />
    <xacro:include filename="$(find velodyne_description)/urdf/VLP-16.urdf.xacro"/>
    <xacro:VLP-16 parent="base_link" name="velodyne" topic="points_raw" organize_cloud="${organize_cloud}" hz="10" samples="440" gpu="${gpu}" lasers="${lasers_horizon}">
        <origin xyz="0 0 0.90315" rpy="0 0 0" />
    </xacro:VLP-16> 

    <!-- 0.8 - 0.03585 =  0.76415-->
    <!-- <xacro:arg name="gpu" default="true"/>
    <xacro:property name="gpu" value="$(arg gpu)" />
    <xacro:arg name="organize_cloud" default="false"/>
    <xacro:property name="organize_cloud" value="$(arg organize_cloud)" />
    <xacro:include filename="$(find velodyne_description)/urdf/VLP-16.urdf.xacro"/>
    <xacro:VLP-16 parent="base_link" name="velodyne1" topic="points_raw1" organize_cloud="${organize_cloud}" hz="10" samples="440" gpu="${gpu}" lasers="${lasers_vertical}">
        <origin xyz="0.2 0 1.06415" rpy="0 1.57 0" />
        
    </xacro:VLP-16>  -->

    <!-- IMU sensor -->
    <joint name="imu_joint1" type="fixed">
        <parent link="base_link"/>
        <child link="imu_link1"/>
        <origin xyz="-0.2 0 0.90415" rpy="0 0 0"/>
    </joint>
    <link name="imu_link1">
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <box size="0.1 0.1 0.1"/>
            </geometry>
        </collision>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <box size="0.1 0.1 0.1"/>
            </geometry>
            <material name="yellow">
                <color rgba="0.5 0.3 0.0 0.5" />
            </material>
        </visual>
    </link>
    <gazebo reference="imu_link1">
        <gravity>true</gravity>
        <material>Gazebo/Yellow</material>
        <sensor name="imu_sensor1" type="imu">
            <always_on>true</always_on>
            <update_rate>200.0</update_rate>
            <visualize>true</visualize>
            <topic>__default_topic__</topic>
            <plugin filename="libgazebo_ros_imu_sensor.so" name="imu_plugin">
                <topicName>/$(arg robot_name)/imu</topicName>
                <frameName>imu_link1</frameName>
                <updateRateHZ>200.0</updateRateHZ>
                <gaussianNoise>0.0</gaussianNoise>
                <xyzOffset>0 0 0</xyzOffset>
                <rpyOffset>0 0 0</rpyOffset>
            </plugin>
            <pose>0 0 0 0 0 0</pose>
        </sensor>
    </gazebo>
    <!--Camera1 sensor-->
    <joint name="camera1_joint" type="fixed">
        <axis xyz="0 1 0" />
        <origin xyz="${base_length/2-0.01} 0 ${base_height/2+base_collision_height+0.01}" rpy="0 0 0"/>
        <parent link="base_link"/>
        <child link="camera1_link"/>
    </joint>
    <link name="camera1_link">
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <box size="0.02 0.02 0.02"/>
            </geometry>
        </collision>
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <box size="0.02 0.02 0.02"/>
            </geometry>
            <!-- <material name="red" color="1 0 0 1"/> -->

            <material name="red">
                <color rgba="1.0 0.0 0.0 1.0"/>
            </material>


        </visual>
    </link>
    <gazebo reference="camera1_link">
        <sensor type="camera" name="camera1">
            <update_rate>30.0</update_rate>
            <camera name="head">
                <horizontal_fov>1.74532925</horizontal_fov>
                <image>
                    <width>800</width>
                    <height>800</height>
                    <format>R8G8B8</format>
                </image>
                <clip>
                    <near>0.02</near>
                    <far>300</far>
                </clip>
                <noise>
                    <type>gaussian</type>
                    <!-- Noise is sampled independently per pixel on each frame.
               That pixel's noise value is added to each of its color
               channels, which at that point lie in the range [0,1]. -->
                    <mean>0.0</mean>
                    <stddev>0.007</stddev>
                </noise>
            </camera>
            <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
                <alwaysOn>true</alwaysOn>
                <updateRate>0.0</updateRate>
                <!-- <cameraName>${robot_name}/camera1</cameraName> -->
                <cameraName>airport_robot/camera1</cameraName>
                <imageTopicName>image_raw</imageTopicName>
                <cameraInfoTopicName>camera_info</cameraInfoTopicName>
                <frameName>camera_link</frameName>
                <hackBaseline>0.07</hackBaseline>
                <distortionK1>0.0</distortionK1>
                <distortionK2>0.0</distortionK2>
                <distortionK3>0.0</distortionK3>
                <distortionT1>0.0</distortionT1>
                <distortionT2>0.0</distortionT2>
            </plugin>
        </sensor>
    </gazebo>

    <!--Rear Right Wheel-->
    <joint name="rear_right_wheel_joint" type="continuous">
        <parent link="base_link"/>
        <child link="rear_right_wheel_link"/>
        <origin xyz="${-rear_tyre_x} ${-rear_tyre_y} ${rear_tyre_r}" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <dynamics damping="0.1"/>
        <limit effort="1000000" velocity="10000" />
        <joint_properties damping="0.0" friction="0.0" />
    </joint>
    <link name="rear_right_wheel_link">
        <collision>
            <origin xyz="0 0 0" rpy="${PI/2} 0 0"/>
            <geometry>
                <cylinder length="${rear_tyre_length}" radius="${rear_tyre_r}"/>
            </geometry>
        </collision>
        <visual>
            <origin xyz="0 0 0" rpy="${PI/2} 0 0"/>
            <geometry>
                <cylinder length="${rear_tyre_length}" radius="${rear_tyre_r}"/>
            </geometry>
        </visual>
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="${rear_tyre_mass}"/>
            <inertia ixx="${rear_tyre_mass_ixx}" ixy="0.000000" ixz="0.000000" iyy="${rear_tyre_mass_iyy}" iyz="0.000000" izz="${rear_tyre_mass_izz}"/>
        </inertial>
    </link>

    <!--Rear Left Wheel-->
    <joint name="rear_left_wheel_joint" type="continuous">
        <parent link="base_link"/>
        <child link="rear_left_wheel_link"/>
        <origin xyz="${-rear_tyre_x} ${rear_tyre_y} ${rear_tyre_r}" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <dynamics damping="0.1"/>
        <limit effort="1000000" velocity="10000" />
        <joint_properties damping="0.0" friction="0.0" />
    </joint>
    <link name="rear_left_wheel_link">
        <collision>
            <origin xyz="0 0 0" rpy="${PI/2} 0 0"/>
            <geometry>
                <cylinder length="${rear_tyre_length}" radius="${rear_tyre_r}"/>
            </geometry>
        </collision>
        <visual>
            <origin xyz="0 0 0" rpy="${PI/2} 0 0"/>
            <geometry>
                <cylinder length="${rear_tyre_length}" radius="${rear_tyre_r}"/>
            </geometry>
        </visual>
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="${rear_tyre_mass}"/>
            <inertia ixx="${rear_tyre_mass_ixx}" ixy="0.000000" ixz="0.000000" iyy="${rear_tyre_mass_iyy}" iyz="0.000000" izz="${rear_tyre_mass_izz}"/>
        </inertial>
    </link>

    <!--Front Right Wheel-->
    <joint name="front_right_wheel_joint" type="continuous">
        <parent link="base_link"/>
        <child link="front_right_wheel_link"/>
        <origin xyz="${front_tyre_x} ${-front_tyre_y} ${front_tyre_r}" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <dynamics damping="0.1"/>
        <limit effort="1000000" velocity="10000" />
        <joint_properties damping="0.0" friction="0.0" />
    </joint>
    <link name="front_right_wheel_link">
        <collision>
            <origin xyz="0 0 0" rpy="${PI/2} 0 0"/>
            <geometry>
                <cylinder length="${front_tyre_length}" radius="${front_tyre_r}"/>
            </geometry>
        </collision>
        <visual>
            <origin xyz="0 0 0" rpy="${PI/2} 0 0"/>
            <geometry>
                <cylinder length="${front_tyre_length}" radius="${front_tyre_r}"/>
            </geometry>
        </visual>
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="${front_tyre_mass}"/>
            <inertia ixx="${front_tyre_mass_ixx}" ixy="0.000000" ixz="0.000000" iyy="${front_tyre_mass_iyy}" iyz="0.000000" izz="${front_tyre_mass_izz}"/>
        </inertial>
    </link>

    <!--Front Left Wheel-->
    <joint name="front_left_wheel_joint" type="continuous">
        <parent link="base_link"/>
        <child link="front_left_wheel_link"/>
        <origin xyz="${front_tyre_x} ${front_tyre_y} ${front_tyre_r}" rpy="0 0 0"/>
        <axis xyz="0 1 0"/>
        <dynamics damping="0.1"/>
        <limit effort="1000000" velocity="10000" />
        <joint_properties damping="0.0" friction="0.0" />

    </joint>
    <link name="front_left_wheel_link">
        <collision>
            <origin xyz="0 0 0" rpy="${PI/2} 0 0"/>
            <geometry>
                <cylinder length="${front_tyre_length}" radius="${front_tyre_r}"/>
            </geometry>
        </collision>
        <visual>
            <origin xyz="0 0 0" rpy="${PI/2} 0 0"/>
            <geometry>
                <cylinder length="${front_tyre_length}" radius="${front_tyre_r}"/>
            </geometry>
        </visual>
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="${front_tyre_mass}"/>
            <inertia ixx="${front_tyre_mass_ixx}" ixy="0.000000" ixz="0.000000" iyy="${front_tyre_mass_iyy}" iyz="0.000000" izz="${front_tyre_mass_izz}"/>
        </inertial>
    </link>

    <!-- motors and transmissions for the two rear wheels -->
    <transmission name="tran1">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="rear_right_wheel_joint">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor1">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="tran2">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="rear_left_wheel_joint">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor2">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <!-- EPS and transmissions for the front steering -->
    <transmission name="tran3">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="front_right_wheel_joint">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor3">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="tran4">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="front_left_wheel_joint">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor4">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    

    <!-- Friction Parametres -->
    <gazebo reference="rear_right_wheel_link">
        <mu1>10000000</mu1>
        <mu2>10000000</mu2>
        <kp>10000000</kp>
        <kd>1</kd>
    </gazebo>
    <gazebo reference="rear_left_wheel_link">
        <mu1>10000000</mu1>
        <mu2>10000000</mu2>
        <kp>10000000</kp>
        <kd>1</kd>
    </gazebo>
    <gazebo reference="front_right_wheel_link">
        <mu1>10000000</mu1>
        <mu2>10000000</mu2>
        <kp>10000000</kp>
        <kd>1</kd>
    </gazebo>
    <gazebo reference="front_left_wheel_link">
        <mu1>10000000</mu1>
        <mu2>10000000</mu2>
        <kp>10000000</kp>
        <kd>1</kd>
    </gazebo>

    <!-- Gazebo Plugins -->
    <gazebo>
        <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
            <robotNamespace>/${robot_name}/airport_robot</robotNamespace>
            <robotParam>robot_description</robotParam>
            <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
        </plugin>
    </gazebo>
    <gazebo>
        <plugin name="mecanum_controller" filename="libgazebo_ros_planar_move.so">
            <robotNamespace>/${robot_name}/airport_robot</robotNamespace>
            <commandTopic>cmd_vel</commandTopic>
            <odometryTopic>odom</odometryTopic>
            <odometryFrame>odom</odometryFrame>
            <odometryRate>1.0</odometryRate>   
            <robotBaseFrame>base_footprint</robotBaseFrame>
            <cmdTimeout>0.5</cmdTimeout>         
        </plugin>
    </gazebo>
</robot>



